src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -16,33 +16,75 @@ open VectorMatrix
 
 module Mpqf = SharedFunctions.Mpqf
 
-module Equality = struct
+module Rhs = struct
   (* (Some i, k) represents a sum of a variable with index i and the number k.
      (None, k) represents the number k. *)
-  type t = (int option * (Z.t [@printer Z.pp_print])) [@@deriving eq, ord, hash, show]
+  type t = (int option * (Z.t [@printer Z.pp_print])) [@@deriving eq, ord, hash]
   let zero = (None, Z.zero)
   let var_zero i = (Some i, Z.zero)
   let to_int x = Z.to_int @@ snd x
+  let show (v, o) = Option.map_default (fun i -> "var_" ^ string_of_int i^ " + ") "" v  ^ Z.to_string o
 end
 
-module EqualitiesArray = struct
-  include Array
-  type t = Equality.t Array.t [@@deriving eq, ord]
+module EqualitiesConjunction = struct
+  module IntHash = struct
+    type t = int
+    let equal = (=)
+    let hash i = i land max_int
+  end
+  module IntHashtbl = BatHashtbl.Make(IntHash)
+
+  let equal i j = (fst i) = (fst j) && IntHashtbl.fold (fun lh rh acc -> acc && IntHashtbl.mem (snd j) lh && IntHashtbl.find (snd j) lh = rh) (snd i) true
+
+  let compare i j = let c = compare (fst i) (fst j) in if c <> 0 then
+      c
+    else let merg = IntHashtbl.merge (fun k a b -> match a,b with
+        | Some a, Some b -> Some (a,b)
+        | Some v, None-> Some (v,Rhs.var_zero k)
+        | None, Some v -> Some (Rhs.var_zero k,v)
+        | None, None -> None) (snd i) (snd j) in
+      let lst= IntHashtbl.to_list merg  in
+      let lst= List.sort (fun (a,_) (b,_) -> compare a b) lst in
+      let rec compare_list li = match li with
+        | [] -> 0
+        | (_,(a,b))::rest -> let c = compare a b in
+          if c = 0 then compare_list rest
+          else c
+      in compare_list lst
+
+  let show econ = let show_rhs i (v, o) = Printf.sprintf "var_%d=%s " i (Rhs.show (v, o)) in
+    IntHashtbl.fold (fun lhs rhs acc -> acc ^ show_rhs lhs rhs) econ ""
+
+  type t = int * Rhs.t IntHashtbl.t
+  let hash : t -> int = fun x -> IntHashtbl.fold (fun k value acc -> 31 * 31 * acc + 31 * k + Rhs.hash value) (snd x) 0 (* TODO: derive *)
+
+  let empty () = (0, IntHashtbl.create 0)
 
-  let show m =
-    Array.fold_right (fun k result -> Equality.show k ^ "\n" ^ result) m ""
+  let make_empty_conj len = (len, IntHashtbl.create len)
 
-  let hash : t -> int = Array.fold_left (fun acc a -> 31 * acc + Equality.hash a) 0 (* TODO: derive *)
+  let get_dim econ = fst econ
 
-  let empty () = [||]
+  let nontrivial econ lhs = IntHashtbl.mem (snd econ) lhs
 
-  let make_empty_array len = Array.init len (fun i -> (Some i, Z.zero))
+  (** sparse implementation of get rhs for lhs, but will default to no mapping for sparse entries *)
+  let get_rhs econ lhs = IntHashtbl.find_default (snd econ) lhs (Rhs.var_zero lhs)
+
+  (** inplace update on the EqualitiesConjunction, repling the rhs for lhs with a new one if meaningful, otherwise just drop the entry*)
```

Comment:
```suggestion
  (** inplace update on the EqualitiesConjunction, replacing the rhs for lhs with a new one if meaningful, otherwise just drop the entry*)
```